### ISSUE TITLE
Presence of several included builds affects build behaviour unexpectedly

### ISSUE THREAD

**Issue by facewindu** [gradle/gradle#4064]
The attached composite build suffers from 2 issues

Issue 1: one of the test task is not executed (and it is, when removing the included builds)
Issue 2: the MultiBuildFailures exceptions doesn't report exceptions from all the included builds.

The attached project is set-up like
-  root build
    - subproject
    - other.subproject
    - nested.build (included build)
      - nested.build.subproject
    - sibling.build (included build)
      - sibling.build.subproject

All those 7 projects are configured to have test execution failures.
The root build contains
```
build.dependsOn gradle.includedBuilds*.task(':build')
build.dependsOn subprojects*.task(':build')
```

### Expected Behavior
Issue 1: 
When running `./gradlew build --continue` from the `root.build`, I expect all 7 test tasks to be executed.

Issue 2:
I also expect the `MultipleBuildFailures` to include 4 causes
- Execution failed for task ':nested.build:test'
- Execution failed for task ':nested.build:nested.build.subproject:test'
- Execution failed for task ':sibling.build:test'
- Execution failed for task ':sibling.build:sibling.build.subproject:test'

### Current Behavior
Issue 1:
`:subproject:test` is currently not executed. If I remove the included builds from the root project `settings.gradle`, it is executed.
Issue 2:
The `MultipleBuildFailures` includes 2 causes only
- Execution failed for task ':nested.build:test'
- Execution failed for task ':nested.build:nested.build.subproject:test'

### Context
The MultipleBuildFailures problem prevents proper integration of composites in build scans, as we extract project failure information from the reported exceptions.

### Steps to Reproduce (for bugs)
Unzip [included-build-issue.zip](https://github.com/gradle/gradle/files/1640659/included-build-issue.zip)
`cd included-build-issue/root.build`
`./gradlew build --continue`
(note that probably using `build --continue --stacktrace` is better given how minimal the info on the multi build failure is otherwise)

### Your Environment
I can provide access to a dev build scan upon request (since it's a composite build and currently no other build scan server / build scan plugin can handle them)

**Comment by ldaley** [comment]
@facewindu it's too late in the release cycle to change this. We are just going to have to wear the cost of adapting on our side.

**Comment by oehme** [comment]
I'm ready to hold up 4.5 if this is a regression in this release. If the problem already existed in 4.4, then I think we should postpone the fix to 4.6.

**Comment by facewindu** [comment]
@oehme I ran it with 4.4.1 and it shows both issues too.

**Comment by bigdaz** [comment]
Regarding "Issue #1", I don't think this line makes sense:
```
build.dependsOn subprojects*.task(':build')
```
This is because `:build` is an _absolute path_, so will always reference the `build` task from the root project. So `subprojects*.task(':build')` is asking for the build task of the root project of each subproject (and the root project is always the same for all subprojects).

It's safe to remove that line from the root `build.gradle` file. The reason that the subproject `build` tasks are being executed at all is because of the command line `./gradlew build` (not `./gradlew :build`).

I think that all reported issues can be explained due to improper handling of `--continue` and parallel-by-default execution in a composite build. So when a failure occurs in an included build task, other tasks that have been scheduled do not get correctly executed, or if executed the failures may not be correctly collected and reported.

See #2520. It seems that support for `--continue` was never fully completed, and got a bit worse when we implemented parallel execution of tasks in a composite.

**Comment by bigdaz** [comment]
I think that these reported issues can be explained due to improper handling of `--continue` and parallel-by-default execution in a composite build. So when a failure occurs in an included build task, other tasks that have been scheduled do not get correctly executed, or if executed the failures may not be correctly collected and reported.

See #2520. It seems that composite-build support for `--continue` was never fully completed, and got a bit worse when we implemented parallel execution of tasks in a composite.

**Comment by ldaley** [comment]
@bigdaz in that case, should this be closed as a duplicate of #2520?

**Comment by oehme** [comment]
I suggest we add this as a limitation to the documentation and add a warning to the console to inform the user that this is not expected to work. Would that be an acceptable outcome from your side as well @ldaley?

**Comment by ldaley** [comment]
WRT build scans, issue 2 is actually the problem. This is because, right now, we use the ultimate build exception to communicate what went wrong to the user. This means that we are not going to show all of the task failures. This might change soon, in that we may use a different approach that doesn't require mining the exception. In such a case, the impact of issue 2 would be far less severe.

However, Gradle not indicating which tasks actually failed is a rather serious issue. I don't feel comfortable not addressing that, but not because of scans. I think we should treat (more or less) silent dropping of failures as a serious bug.

I have no specific issue with treating `--continue` with composites as unreliable (i.e. issue 1).

**Comment by oehme** [comment]
But if we tell the user that `--continue` won't work, then issue 2) would no longer happen, right? I mean isn't `--continue` the only time when we have multiple failures?

**Comment by ldaley** [comment]
Tasks running concurrently may fail, and the build boundary is implicitly a concurrency boundary, so —continue is not the only case.

**Comment by oehme** [comment]
Got it, so this is separate from #2520.

**Comment by bmuschko** [comment]
> Tasks running concurrently may fail, and the build boundary is implicitly a concurrency boundary, so `--continue` is not the only case.

Just to verify - we have the issue when running with `--parallel` and/or `--continue`. Correct? Are there are other flags that need to be taken under consideration?

**Comment by bigdaz** [comment]
No. `--parallel` is not involved, and I don't think `--continue` isn't required for us to miss collecting all of the failures.

We _could_ add a warning about using `--continue` with composite builds, since it probably makes things worse.

- _Without_ `--continue` (and without `--parallel`), we will execute multiple builds in parallel, and only report failures from the first one that fails.
- _With_ `--continue` (and without `--parallel`), we will execute multiple builds in parallel, and only report failures from the first one that fails.

So it's the same behaviour in both cases, but with `--continue`, users _expect_ to see _all_ of the failures. Without `--continue` they are used to only seeing the first failure that occurs.

**Comment by ldaley** [comment]
> So it's the same behaviour in both cases, but with --continue, users expect to see all of the failures. 

I see your point regarding expectation, but just wanted to point out for completeness that this doesn't quite align with the non composite `--parallel` case. There we will report multiple errors if concurrent tasks failed. I don't think it should be any different for composite builds (i.e. all failures should be surfaced), and you can make a weaker expectation argument that users (based on their experience with `--parallel`) expect all failures to surface.

**Comment by bmuschko** [comment]
PRs:
- https://github.com/gradle/gradle/pull/4295
- https://github.com/gradle/gradle/pull/4324

**Comment by oehme** [comment]
@bmuschko with #4324 merged, is #4295 still necessary?

**Comment by breskeby** [comment]
yeah, the message provided by #4295 is misleading as all failures are reported as I would have expected

**Comment by bmuschko** [comment]
@oehme @breskeby Regarding https://github.com/gradle/gradle/pull/4295 there might be other limitations. It would be best if @bigdaz could comment on that. I don't have the full picture.

**Comment by bigdaz** [comment]
@breskeby Thanks for reporting that issue. It's actually not related to the use of `--continue`, but is due to a single included build providing multiple artifacts in a composite. 

I've raised #4370 for this issue. We already had a test case marked `@ToBeImplemented` demonstrating the issue, but it's good to have a clear bug report for this.

**Comment by bmuschko** [comment]
@breskeby I remove the warning message on `release` (see https://github.com/gradle/gradle/pull/4372).

**Comment by breskeby** [comment]
@bigdaz okay good to know. I added a workaround for the duplicate reporting in the build scan plugin.
